From 2d6c7b7cca14b1a314da21cbb58c044ce2ab6f5b Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:13:51 -0400
Subject: [powerpc] use _GLOBAL_TOC for memmove

Message-id: <1410545655-205645-603-git-send-email-dzickus@redhat.com>
Patchwork-id: 94483
O-Subject: [RHEL7 PATCH 602/626] powerpc: use _GLOBAL_TOC for memmove
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit 6f5405bc2ee0102bb3856e2cdea64ff415db2e0c
Author: Li Zhong <zhong@linux.vnet.ibm.com>
Date:   Mon Jul 21 17:55:13 2014 +0800

    powerpc: use _GLOBAL_TOC for memmove

    memmove may be called from module code copy_pages(btrfs), and it may
    call memcpy, which may call back to C code, so it needs to use
    _GLOBAL_TOC to set up r2 correctly.

    This fixes following error when I tried to boot an le guest:

    Vector: 300 (Data Access) at [c000000073f97210]
        pc: c000000000015004: enable_kernel_altivec+0x24/0x80
        lr: c000000000058fbc: enter_vmx_copy+0x3c/0x60
        sp: c000000073f97490
       msr: 8000000002009033
       dar: d000000001d50170
     dsisr: 40000000
      current = 0xc0000000734c0000
      paca    = 0xc00000000fff0000  softe: 0  irq_happened: 0x01
        pid   = 815, comm = mktemp
    enter ? for help
    [c000000073f974f0] c000000000058fbc enter_vmx_copy+0x3c/0x60
    [c000000073f97510] c000000000057d34 memcpy_power7+0x274/0x840
    [c000000073f97610] d000000001c3179c copy_pages+0xfc/0x110 [btrfs]
    [c000000073f97660] d000000001c3c248 memcpy_extent_buffer+0xe8/0x160 [btrfs]
    [c000000073f97700] d000000001be4be8 setup_items_for_insert+0x208/0x4a0 [btrfs]
    [c000000073f97820] d000000001be50b4 btrfs_insert_empty_items+0xf4/0x140 [btrfs]
    [c000000073f97890] d000000001bfed30 insert_with_overflow+0x70/0x180 [btrfs]
    [c000000073f97900] d000000001bff174 btrfs_insert_dir_item+0x114/0x2f0 [btrfs]
    [c000000073f979a0] d000000001c1f92c btrfs_add_link+0x10c/0x370 [btrfs]
    [c000000073f97a40] d000000001c20e94 btrfs_create+0x204/0x270 [btrfs]
    [c000000073f97b00] c00000000026d438 vfs_create+0x178/0x210
    [c000000073f97b50] c000000000270a70 do_last+0x9f0/0xe90
    [c000000073f97c20] c000000000271010 path_openat+0x100/0x810
    [c000000073f97ce0] c000000000272ea8 do_filp_open+0x58/0xd0
    [c000000073f97dc0] c00000000025ade8 do_sys_open+0x1b8/0x300
    [c000000073f97e30] c00000000000a008 syscall_exit+0x0/0x7c

    Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/lib/mem_64.S b/arch/powerpc/lib/mem_64.S
index 0738f96..43435c6 100644
--- a/arch/powerpc/lib/mem_64.S
+++ b/arch/powerpc/lib/mem_64.S
@@ -77,7 +77,7 @@ _GLOBAL(memset)
  stb r4,0(r6)
  blr
 
-_GLOBAL(memmove)
+_GLOBAL_TOC(memmove)
  cmplw 0,r3,r4
  bgt backwards_memcpy
  b memcpy
-- 
1.7.1